### PR TITLE
[Editor, Servers] Minor optimizations

### PR DESCRIPTION
**Pull request by stuartcarnie**
The PR introduces a few minor optimisations to prevent double look-ups in hash tables and removes some redundant checks.

### CODE REVIEW

File: editor/editor_inspector.cpp [hunk 1]
```diff
@@ -2246,7 +2246,7 @@ void EditorInspectorArray::_setup() {
 			}
 			move_vbox->add_child(ae.move_texture_rect);
 
-			if (element_position < _get_array_count() - 1) {
+			if (element_position < count - 1) {
```
Comment (stuartcarnie) [inline editor/editor_inspector.cpp:2249]: Counts are cached at the top of the function, so no need to call this again here.

File: servers/audio_server.cpp [hunk 2]
```diff
@@ -1487,7 +1487,10 @@ void AudioServer::init() {
 
 void AudioServer::update() {
 #ifdef DEBUG_ENABLED
-	if (EngineDebugger::is_profiling("servers")) {
+	// this prevents allocations every update, as "servers" does not exist
+	// 4.57 Mc    0.0%    1.40 Mc               AudioServer::update()
+	static StringName const SERVERS("servers", true);
+	if (EngineDebugger::is_profiling(SERVERS)) {
```
Comment (stuartcarnie) [inline servers/audio_server.cpp:1493]: I'm not sure what the correct fix is here, but this causes repeated allocations in the game loop.

File: servers/rendering/renderer_viewport.cpp
Comment (stuartcarnie) [inline servers/rendering/renderer_viewport.cpp:1]: Only perform a single lookup in the hash table

File: servers/rendering/rendering_device.cpp [hunk 3]
```diff
@@ -138,17 +138,17 @@ RenderingDevice::ShaderSPIRVGetCacheKeyFunction RenderingDevice::get_spirv_cache
...
-
-	if (!reverse_dependency_map.has(p_id)) {
-		reverse_dependency_map[p_id] = HashSet<RID>();
+	set = reverse_dependency_map.getptr(p_id);
+	if (set == nullptr) {
+		set = &reverse_dependency_map.insert(p_id, HashSet<RID>())->value;
 	}
-
-	reverse_dependency_map[p_id].insert(p_depends_on);
+	set->insert(p_depends_on);
```
Comment (stuartcarnie) [inline servers/rendering/rendering_device.cpp:151]: Previously this code was always performing 2 hash table lookups, one via `has` and then a second via `operator[]`. Changed to cache the lookup once via a pointer.

File: servers/rendering/rendering_server_default.cpp [hunk 4]
```diff
@@ -150,12 +150,10 @@ void RenderingServerDefault::_draw(bool p_swap_buffers, double frame_step) {
...
-			if (name[0] != '<' && name[0] != '>') {
-				if (print_gpu_profile_task_time.has(name)) {
-					print_gpu_profile_task_time[name] += time;
-				} else {
-					print_gpu_profile_task_time[name] = time;
-				}
+			if (print_gpu_profile_task_time.has(name)) {
+				print_gpu_profile_task_time[name] += time;
+			} else {
+				print_gpu_profile_task_time[name] = time;
```
Comment (stuartcarnie) [inline servers/rendering/rendering_server_default.cpp:156]: This if was redundant, as the loop already checks if `name[0] == '<' || name[0] == '>'` and continues.

File: servers/rendering/rendering_server_default.cpp [hunk 5]
```diff
@@ -150,12 +150,10 @@ void RenderingServerDefault::_draw(bool p_swap_buffers, double frame_step) {
 
 			double time = frame_profile[i + 1].gpu_msec - frame_profile[i].gpu_msec;
 
-			if (name[0] != '<' && name[0] != '>') {
```
Comment (stuartcarnie) [inline servers/rendering/rendering_server_default.cpp:153]: This if was redundant, as the loop already checks if `name[0] == '<' || name[0] == '>'` and continues.

File: servers/rendering/rendering_device.cpp [hunk 6]
```diff
@@ -138,17 +138,17 @@ RenderingDevice::ShaderSPIRVGetCacheKeyFunction RenderingDevice::get_spirv_cache
...
 void RenderingDevice::_add_dependency(RID p_id, RID p_depends_on) {
-	if (!dependency_map.has(p_depends_on)) {
-		dependency_map[p_depends_on] = HashSet<RID>();
+	HashSet<RID> *set = dependency_map.getptr(p_depends_on);
+	if (set == nullptr) {
+		set = &dependency_map.insert(p_depends_on, HashSet<RID>())->value;
 	}
+	set->insert(p_id);
 
-	dependency_map[p_depends_on].insert(p_id);
```
Comment (stuartcarnie) [inline servers/rendering/rendering_device.cpp:145]: The improvement removes a hash table lookup, as it calls `getptr` to the value and only if it is `nullptr`, does it insert it. Overall this function performs 2 less hash-table lookups per call.

File: servers/audio_server.cpp (hunk 2 above)
Comment (clayjohn) [reply servers/audio_server.cpp:1493]: For this we usually allocate a static StringName. For example for the scene classes we use:
https://github.com/godotengine/godot/blob/2c136e6170a40f58f2dfb89d32eadfca7156ef37/scene/scene_string_names.cpp#L35

We also have a CoreStringNames class and an EditorStringNames class. Perhaps its time to create ServerStringNames?

Comment (clayjohn) [reply servers/audio_server.cpp:1493]: CC @KoBeWi

Comment (stuartcarnie) [reply servers/audio_server.cpp:1493]: Cool, I thought it might have been something like that, as I'd seen them elsewhere 👍🏻

Comment (clayjohn) [reply servers/audio_server.cpp:1493]: Actually should probably just go in core_string_names.cpp. That seems to be included everywhere

Comment (KoBeWi) [reply servers/audio_server.cpp:1493]: You can just use SNAME.

Comment (stuartcarnie) [reply servers/audio_server.cpp:1493]: I can remove this change from the PR, if the other changes are interesting.

Comment (stuartcarnie) [reply servers/audio_server.cpp]: Reverted this change from the PR

Comment (stuartcarnie) [reply servers/audio_server.cpp]: Oh nice – I'll make that change then 👍🏻

Comment (stuartcarnie) [reply servers/audio_server.cpp]: Done

**Review by clayjohn — APPROVED** [review]
Looks fine to me!

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**